### PR TITLE
Review and improvements in databases drivers MySQLi, Postgre and SQLite:

### PR DESCRIPTION
**Pull request by fmertins**
1) Fixed typo method name "escapeLikeStr()" to "escapeLikeString()".
2) Methods _fieldData(), _indexData() and _foreignKeyData() now return only array and throw DatabaseException (instead of returning false).
2.1) BaseConnection, added abstract methods() missing because this class relies on all its subclasses having those methods.
3) BaseBuilder, added TODO about a parameter with default value.
4) composer.json, added ext-mysqli at require-dev section.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -24,7 +24,8 @@
     "require-dev": {
         "phpunit/phpunit": "^7.0",
         "mikey179/vfsStream": "1.6.*",
-        "codeigniter4/codeigniter4-standard": "^1.0"
+        "codeigniter4/codeigniter4-standard": "^1.0",
+        "ext-mysqli": "*"
```
Comment (lonnieezell) [inline composer.json]: Mysql driver is not needed if you plan on using Postgres. Please remove.

File: system/Database/MySQLi/Connection.php [hunk 2]
```diff
@@ -408,16 +408,19 @@ protected function _listColumns(string $table = ''): string
...
-	public function _fieldData(string $table)
+	public function _fieldData(string $table): array
 	{
-		if (($query = $this->query('SHOW COLUMNS FROM ' . $this->protectIdentifiers($table, TRUE, NULL, FALSE))) === FALSE)
+		$table = $this->protectIdentifiers($table, TRUE, NULL, FALSE);
+
+		if (($query = $this->query('SHOW COLUMNS FROM ' . $table)) === FALSE)
 		{
-			return FALSE;
+			throw new DatabaseException('Failed to get field data from MySQL.');
```
Comment (lonnieezell) [inline system/Database/MySQLi/Connection.php]: Please use a localized string that is added to `system/Language/en/Database.php`

File: system/Database/MySQLi/Connection.php [hunk 3]
```diff
@@ -440,21 +442,25 @@ public function _fieldData(string $table)
...
-	public function _indexData(string $table)
+	public function _indexData(string $table): array
 	{
-		if (($query = $this->query('SHOW CREATE TABLE ' . $this->protectIdentifiers($table, TRUE, NULL, FALSE))) === FALSE)
+		$table = $this->protectIdentifiers($table, TRUE, NULL, FALSE);
+
+		if (($query = $this->query('SHOW CREATE TABLE ' . $table)) === FALSE)
 		{
-			return FALSE;
+			throw new DatabaseException('Failed to get index data from MySQL.');
```
Comment (lonnieezell) [inline system/Database/MySQLi/Connection.php]: Use localized string here.

File: system/Database/MySQLi/Connection.php [hunk 4]
```diff
@@ -488,7 +494,7 @@ public function _indexData(string $table)
 				}
 				else
 				{
-					throw new \LogicException('parsing key string failed.');
+					throw new \LogicException('Parsing key string failed.');
```
Comment (lonnieezell) [inline system/Database/MySQLi/Connection.php]: If you wouldn't mind moving this one to the Database lang file too, that would be great!

File: system/Database/MySQLi/Connection.php [hunk 5]
```diff
@@ -521,7 +528,7 @@ public function _foreignKeyData(string $table)
 
 		if (($query = $this->query($sql)) === false)
 		{
-			return false;
+			throw new DatabaseException('Failed to get foreign key data from MySQL');
```
Comment (lonnieezell) [inline system/Database/MySQLi/Connection.php]: Localize this string

File: system/Database/Postgre/Connection.php [hunk 6]
```diff
@@ -304,7 +306,7 @@ public function _fieldData(string $table)
 
 		if (($query = $this->query($sql)) === false)
 		{
-			return false;
+			throw new DatabaseException('Failed to get field data from PostgreSQL.');
```
Comment (lonnieezell) [inline system/Database/Postgre/Connection.php]: Localize this string

File: system/Database/Postgre/Connection.php [hunk 7]
```diff
@@ -338,7 +341,7 @@ public function _indexData(string $table)
 
 		if (($query = $this->query($sql)) === false)
 		{
-			return false;
+			throw new DatabaseException('Failed to get index data from PostgreSQL.');
```
Comment (lonnieezell) [inline system/Database/Postgre/Connection.php]: Localize this string

File: system/Database/Postgre/Connection.php [hunk 8]
```diff
@@ -386,11 +390,12 @@ public function _foreignKeyData(string $table)
...
                         JOIN information_schema.constraint_column_usage AS ccu
                             ON ccu.constraint_name = tc.constraint_name
-                        WHERE constraint_type = '.$this->escape('FOREIGN KEY').' AND tc.table_name = '.$this->escape($table);
+                        WHERE constraint_type = '.$this->escape('FOREIGN KEY').' AND
+                            tc.table_name = '.$this->escape($table);
 
 		if (($query = $this->query($sql)) === false)
 		{
-			return false;
+			throw new DatabaseException('Failed to get foreign key data from PostgreSQL.');
```
Comment (lonnieezell) [inline system/Database/Postgre/Connection.php]: Localize this string

File: system/Database/SQLite3/Connection.php [hunk 9]
```diff
@@ -291,24 +291,23 @@ public function getFieldNames($table)
...
 	 */
-	public function _fieldData(string $table)
+	public function _fieldData(string $table): array
 	{
-
 		if (($query = $this->query('PRAGMA TABLE_INFO('.$this->protectIdentifiers($table, true, null,
 					false).')')) === false)
 		{
-			return false;
+			throw new DatabaseException('Failed to get field data from SQLite.');
```
Comment (lonnieezell) [inline system/Database/SQLite3/Connection.php]: Localize this string

File: system/Database/SQLite3/Connection.php [hunk 10]
```diff
@@ -327,20 +326,20 @@ public function _fieldData(string $table)
...
+	public function _indexData(string $table): array
 	{
 		// Get indexes
 		// Don't use PRAGMA index_list, so we can preserve index order
-		$sql = "SELECT name FROM sqlite_master WHERE type='index' AND tbl_name=".$this->escape(strtolower($table))."";
+		$sql = "SELECT name FROM sqlite_master WHERE type='index' AND tbl_name=".$this->escape(strtolower($table));
 		if (($query = $this->query($sql)) === false)
 		{
-			return false;
+			throw new DatabaseException('Failed to get index data from SQLite.');
```
Comment (lonnieezell) [inline system/Database/SQLite3/Connection.php]: Localize this string

File: system/Database/SQLite3/Connection.php [hunk 11]
```diff
@@ -354,7 +353,7 @@ public function _indexData(string $table)
 			$obj->fields = [];
 			if (($fields = $this->query('PRAGMA index_info('.$this->escape(strtolower($row->name)).')')) === false)
 			{
-				return false;
+				throw new DatabaseException('Failed to get index_info() from SQLite.');
```
Comment (lonnieezell) [inline system/Database/SQLite3/Connection.php]: Localize this string

**Comment by fmertins** [comment]
Hi Lonnie thank you I'll do these improvements still this week.

**Comment by lonnieezell** [comment]
Thanks!

**Pull request merged**